proposals/rich-content-graphql-api.md

```diff
@@ -0,0 +1,331 @@
+            - `PageBuilder_Row`
+            - `PageBuilder_Text`
+
+            (The bundle system and service worker will load from their caches if possible.)
+
+        2.  Run a subsequent query provided by the components' respective implementations:
+
+            ```graphql
+            fragment PageBuilder_Column_Config on PageBuilder_Column {
+              appearance
+              backgroundColor
+              backgroundImage
+              backgroundPosition
+              backgroundSize
+              backgroundRepeat
+              backgroundAttachment
+              minHeight
+              verticalAlign
+            }
+
+            fragment PageBuilder_Banner_Config on PageBuilder_Banner {
+              appearance
+              backgroundType
+              "More banner config here."
+            }
+
+            """
+            Additional fragments here.
+            """
+
+            content {
+              ...PageBuilder_Column_Config
+              children {
+                ... on PageBuilder_Banner {
+                  ...PageBuilder_Banner_Config
+                }
+                ... on PageBuilder_Accordion {
+            	  ...PageBuilder_Accordion_Config
+                }
+                ... on PageBuilder_Text {
+                  ...PageBuilder_Text_Config
+                }
+            	children {
+                  ...NodeMetadata
+                }
+              }
+            }
+            ```
+
+        3.  Re-render the fully hydrated second level of component when that query returns results.
+
+Subsequent requests for any content which lists any of those components in its `descendentRoles` will automatically include those fragment and clauses in the `content` query. Therefore, subsequent requests will contain the necessary configurations on the first pass.
+
+#### Leaf Node
+
+A "leaf node" Slider widget might declare these schema additions:
+
+```graphql
+type SliderSlide {
+  alt: String!
+  src: String!
+}
+
+type PageBuilder_Slider implements VisualContentNode {
+  role: String!
+  depth: Int
+  preload: [String]
+  descendentRoles: [String]
+  children: [VisualContentNode]
+  html: String
+  autoStart: Boolean
+  slides: [SliderSlide]
+}]]>
+```
+
+A particular instance of `Slider` in response to a query for its ID against the `VisualContentNode` interface would look like:
+
+```json
+{
+  "content": {
+    "role": "Slider",
+    "preload": ["slide1.jpg", "slide2.jpg", "slide3.jpg"],
+    "depth": 0,
+    "descendentRoles": [],
+    "children": [],
+    "html": ""
+  }
+}
+```
+
+The PWA would handle this response:
+
+1.  In parallel:
+
+    1.  Render the `html` content by injecting it into a `RawHTML` component.
+
+    2.  Initiate preload requests for all the URLs in the `preload` field.
+
+    3.  In series:
+
+    4.  Dynamically load and inject the `PageBuilder_Slider` React Component. (The bundle system and service worker will load from their caches if possible.)
```

Comment:
How much content are we talking about here? I'm concerned about performance issues of putting too much responsibility on React to render things on the FE instead of on the first page load. I only speak from experience, because I've been 🔥  before on this point.

Reply:
For Slider, we're talking about 3-4 Slides usually, with something like 10-20 is going to be a very rare edge case. 

But the thing is is that the Slide itself can contain more internal components, like buttons, headings, texts, etc.

Reply:
I think we'd have to see how it performed. The HTML fallback would already be in the DOM node that the Slider component takes over, so I would hope the React reconciler would reuse as much of it as possible.

How much content are we talking? The point is, we don't know at runtime.